File: 1046-last-stone-weight/1046-last-stone-weight.java
class Solution {
    public int lastStoneWeight(int[] stones) {
        /*Time Complexity :- BigO(NlogN)
         Space Complexity :- BigO(N)*/
         PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for(int s : stones)
            pq.offer(s);
        
        while(pq.size() > 1){
            int x= pq.poll();
            int y=pq.poll();
            if(x!=y) pq.offer(x-y);
           
        }
        if(pq.size() == 1)return pq.poll();
            else return 0;
    }
}

File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        //default max heap 
        int stonessz=stones.size();
        priority_queue<int>pq;
        
        for(int i=0;i<stonessz;i++)
            pq.push(stones[i]);
        
        while(pq.size() > 1)
        {
            int x=pq.top(); pq.pop();
            int y=pq.top(); pq.pop();
            if(x>y)
                pq.push(x-y);
        }
        
        if(pq.size() == 1) return pq.top();
        else return 0;
    }
};